### PR TITLE
Apiverv2

### PR DESCRIPTION
**Pull request by mpnowacki-reef**

### CODE REVIEW

File: b2/json_encoder.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ class B2CliJsonEncoder(json.JSONEncoder):
     def default(self, obj):
         if isinstance(obj, set):
             return list(obj)
-        elif isinstance(obj, FileVersionInfo):
```
Comment (mszumocki-reef) [inline b2/json_encoder.py:29]: merge those ifinstance checks to one for cases when you're doing as_dict() anyway?

File: doc/source/conf.py [hunk 2]
```diff
@@ -196,3 +199,19 @@
...
+white_spaces_start = re.compile(r'^\s*')
+
+
+def setup(_):
+    main_help_text = str(B2.lazy_get_description())
+    indentation = min(len(white_spaces_start.match(line).group(0)) for line in main_help_text.splitlines() if len(line) > 1)
+    main_help_text = '\n'.join(line[indentation:] for line in main_help_text.splitlines())
+    main_help_path = path.join(path.dirname(__file__), 'main_help.rst')
+    if path.exists(main_help_path):
+        with open(main_help_path, 'r') as main_help_file:
```
Comment (mszumocki-reef) [inline doc/source/conf.py:213]: Why open the file twice? You could just change the mode to 'rw'

File: test/integration/test_b2_command_line.py [hunk 3]
```diff
@@ -661,11 +679,11 @@ def account_test(b2_tool, bucket_name):
             b2_tool.application_key,
         ]
     )
-    tearDown_envvar_test('B2_ACCOUNT_INFO')
+    b2_tool.tearDown_envvar_test('B2_ACCOUNT_INFO')
```
Comment (mszumocki-reef) [inline test/integration/test_b2_command_line.py:682]: not even try: finally: here?

File: test/unit/test_console_tool.py [hunk 4]
```diff
@@ -30,9 +35,91 @@ def file_mod_time_millis(path):
...
+        actual_stderr = self._trim_trailing_spaces(stderr.getvalue())
+
+        if '' != actual_stderr:
+            print('ACTUAL STDERR:  ', repr(actual_stderr))
+            print(actual_stderr)
+
+        self.assertEqual('', actual_stderr, 'stderr')
+        self.assertEqual(0, actual_status, 'exit status code')
+
+    def _trim_leading_spaces(self, s):
```
Comment (mszumocki-reef) [inline test/unit/test_console_tool.py:72]: This whole function looks like it could be replaced with call to textwrap.dedent

**Review by mszumocki-reef — CHANGES_REQUESTED** [review]

File: doc/source/conf.py (hunk 2 above)
Comment (mpnowacki-reef) [reply doc/source/conf.py:213]: opening the file with `w+` changes it modification time which causes `nox -s doc` to rebuild the docs and fall into an endless loop

Comment (mszumocki-reef) [reply doc/source/conf.py:213]: w+ always truncates the file. r+ is what you need

Comment (mpnowacki-reef) [reply doc/source/conf.py:213]: `r+` doesn't work if the file does not exist

**Review by mszumocki-reef — APPROVED** [review]

**Review by mszumocki-reef — APPROVED** [review]

File: test/unit/test_console_tool.py [hunk 5]
```diff
@@ -30,9 +35,91 @@ def file_mod_time_millis(path):
...
+    def _run_command_ignore_output(self, argv):
+        """
+        Runs the given command in the console tool, checking that it
+        success, but ignoring the stdout.
+        """
+        stdout, stderr = self._get_stdouterr()
+        actual_status = ConsoleTool(self.b2_api, stdout, stderr).run_command(['b2'] + argv)
+        actual_stderr = self._trim_trailing_spaces(stderr.getvalue())
+
+        if '' != actual_stderr:
```
Comment (ppolewicz) [inline test/unit/test_console_tool.py:65]: Should be variable vs hardcoded string. Other way not. Weird it sounds. Yoda like.

File: test/unit/test_console_tool.py [hunk 6]
```diff
@@ -47,13 +134,89 @@ def assertListOfDictsIsContained(self, list_of_subsets, list_of_supersets):
...
-    def setUp(self):
-        self.account_info = StubAccountInfo()
-        self.cache = None
-        self.raw_api = RawSimulator()
-        self.b2_api = B2Api(self.account_info, self.cache, self.raw_api)
-        (self.account_id, self.master_key) = self.raw_api.create_account()
+    def _authorize_account(self):
+        """
+        Prepare for a test by authorizing an account and getting an
+        account auth token
```
Comment (ppolewicz) [inline test/unit/test_console_tool.py:140]:
```suggestion
        Prepare for a test by authorizing an account and getting an account auth token
```

File: test/integration/test_b2_command_line.py [hunk 7]
```diff
@@ -1398,13 +1411,18 @@ def sse_b2_test(b2_tool, bucket_name):
...
-        ['download-file-by-name', '--noProgress', bucket_name, 'encrypted', os.devnull]
-    )
-    b2_tool.should_succeed(
-        ['download-file-by-name', '--noProgress', bucket_name, 'not_encrypted', os.devnull]
-    )
+    with TempDir() as dir_path:
+        p = lambda fname: os.path.join(dir_path, fname)
+        b2_tool.should_succeed(
+            ['download-file-by-name', '--noProgress', bucket_name, 'encrypted',
+             p('encrypted')]
```
[the remark this answers is not in the document]
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:1457]: right

[the remark this answers is not in the document]
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:1457]: done https://github.com/Backblaze/B2_Command_Line_Tool/issues/733

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,12 +9,31 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 * Fix integration test cleanup
+* Representing encryption-related metadata in buckets and file versions is now consistent
+
+### Changed
+* CLI now uses `b2sdk.v2`
+* Downloading files prints file metadata as soon as the download commences (not when it finishes)
+* New way of establishing location of the SQLite cache file, using `XDG_CONFIG_HOME` env var
+* Downloaded file's metadata is complete and is displayed before the file is downloaded
+* `contentLength` changed to `size` where appropriate
+* Log configuration: stack traces are not printed in case of errors byt default, `--verbose` changes that
```
Comment (ppolewicz) [inline CHANGELOG.md:29]: s/byt/by/

File: CHANGELOG.md [hunk 9]
```diff
@@ -9,12 +9,31 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+### Changed
+* CLI now uses `b2sdk.v2`
+* Downloading files prints file metadata as soon as the download commences (not when it finishes)
+* New way of establishing location of the SQLite cache file, using `XDG_CONFIG_HOME` env var
+* Downloaded file's metadata is complete and is displayed before the file is downloaded
+* `contentLength` changed to `size` where appropriate
+* Log configuration: stack traces are not printed in case of errors byt default, `--verbose` changes that 
+* Log configuration arguments bahaviour altered: `--logConfig` is exclusive with `--verbose` and `--debugLogs`
+* Log configuration arguments bahaviour altered: `--verbose` and `--debugLogs` can be used at the some time
```
Comment (ppolewicz) [inline CHANGELOG.md:31]: s/some/same/

File: CHANGELOG.md [hunk 10]
```diff
@@ -9,12 +9,31 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+* New way of establishing location of the SQLite cache file, using `XDG_CONFIG_HOME` env var
+* Downloaded file's metadata is complete and is displayed before the file is downloaded
+* `contentLength` changed to `size` where appropriate
+* Log configuration: stack traces are not printed in case of errors byt default, `--verbose` changes that 
+* Log configuration arguments bahaviour altered: `--logConfig` is exclusive with `--verbose` and `--debugLogs`
+* Log configuration arguments bahaviour altered: `--verbose` and `--debugLogs` can be used at the some time 
+  (and they will both be taken into account)
+
+### Removed
+* Support of `--metadataDirective` argument in `copy-file-by-id`
```
Comment (ppolewicz) [inline CHANGELOG.md:35]: + `(it's detected automatically)`

File: b2/console_tool.py [hunk 11]
```diff
@@ -1012,12 +1105,13 @@ def _setup_parser(cls, parser):
...
         encryption_setting = self._get_source_sse_setting(args)
-        self.api.download_file_by_id(
-            args.fileId, download_dest, progress_listener, encryption=encryption_setting
+        downloaded_file = self.api.download_file_by_id(
+            args.fileId, progress_listener, encryption=encryption_setting
         )
-        self.console_tool._print_download_info(download_dest)
+        self.console_tool._print_download_info(downloaded_file)
+        downloaded_file.save_to(args.localFileName)
+        self._print('Download finished')
```
Comment (ppolewicz) [inline b2/console_tool.py:1114]: this should be added to changelog

File: b2/console_tool.py [hunk 12]
```diff
@@ -2165,37 +2274,101 @@ def _print(self, *args, **kwargs):
...
             formatter.converter = time.gmtime
             handler = logging.FileHandler('b2_cli.log')
-            handler.setLevel(logging.DEBUG)
             handler.setFormatter(formatter)
 
-            b2_logger = logging.getLogger('b2')
-            b2_logger.setLevel(logging.DEBUG)
-            b2_logger.addHandler(handler)
-            b2_logger.propagate = False
+            logger.setLevel(logging.DEBUG)
```
Comment (ppolewicz) [inline b2/console_tool.py:2370]: you've already set this to debug in line 2354

**Pull request merged**